File: 2023/day2/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

/*
 * =========================================================
 *                        CONSTANTS
 * =========================================================
 */
const MAXRED = 12
const MAXGREEN = 13
const MAXBLUE = 14

/*
 * =========================================================
 *                         UTILITY
 * =========================================================
 */
func readFile(filename string) []string {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Println("Error opening file")
	}

	fileScanner := bufio.NewScanner(file)
	fileScanner.Split(bufio.ScanLines)

	var lines []string
	for fileScanner.Scan() {
		lines = append(lines, fileScanner.Text())
	}
	file.Close()

	return lines
}

/*
 * =========================================================
 *                       SOLUTIONS
 * =========================================================
 */
func sln_one() int {
	lines := readFile("problem.txt")
	var sum int

	// iterating over the games
	for idx, line := range lines {
		var red, green, blue int

		// parsing into a list of cubes
		game := strings.Split(line, ": ")[1]
		game = strings.Replace(game, ", ", ",", -1)
		game = strings.Replace(game, "; ", ",", -1)

		// iterating over the cubes
		for _, cube := range strings.Split(game, ",") {
			subs := strings.Split(cube, " ")
			num, _ := strconv.Atoi(subs[0])
			color := subs[1]

			switch color {
			case "red":
				if num > red {
					red = num
				}
			case "green":
				if num > green {
					green = num
				}
			case "blue":
				if num > blue {
					blue = num
				}
			default:
				fmt.Println("[X] Error: unknown color")
			}
		}

		// check if any of the cube set goes over the ceiling
		if red > MAXRED || green > MAXGREEN || blue > MAXBLUE {
			continue // skip
		} else {
			sum += idx + 1
		}
	}

	return sum
}

func sln_two() int {
	lines := readFile("problem.txt")
	var sum int

	// iterating over the games
	for _, line := range lines {
		var red, green, blue int
		game := strings.Split(line, ": ")[1]
		game = strings.Replace(game, ", ", ",", -1)
		game = strings.Replace(game, "; ", ",", -1)

		// iterating over the cubes
		for _, cube := range strings.Split(game, ",") {
			subs := strings.Split(cube, " ")
			num, _ := strconv.Atoi(subs[0])
			color := subs[1]

			switch color {
			case "red":
				if num > red {
					red = num
				}
			case "green":
				if num > green {
					green = num
				}
			case "blue":
				if num > blue {
					blue = num
				}
			default:
				fmt.Println("[X] Error: unknown color")
			}
		}

		sum += red * green * blue
	}

	return sum
}

func main() {
	var sln int

	sln = sln_one()
	fmt.Printf("[+] part-1 solution: %v\n", sln) // 2727

	sln = sln_two()
	fmt.Printf("[+] part-2 solution: %v\n", sln) // 56580
}

File: 2023/day2/solution.py
#!usr/bin/env python3

FILE = open('problem.txt', 'r').read()
MAXRED = 12
MAXGREEN = 13
MAXBLUE = 14

def sln_one():
    lines = FILE.split('\n')
    sum = 0

    # enumerate each line
    for i, line in enumerate(lines):
        cubes = {
            'red': 0,
            'green': 0,
            'blue': 0
        }

        game, set = line.split(":")
        set = [s.strip() for s in set.replace(";", ",").split(",")] # convert sets of game into list of cubes
        
        # enumerate each cubes
        for s in set:
            num, cube = s.split(" ")
            if(int(num) > cubes[cube]):
                cubes[cube] = int(num)

        if(cubes['red'] <= MAXRED and cubes['green'] <= MAXGREEN and cubes['blue'] <= MAXBLUE):
            sum += i+1

    print(f'[+] part-1 solution: {sum}')

def sln_two():
    lines = FILE.split('\n')
    sum = 0

    # enumerate each line
    for i, line in enumerate(lines):
        cubes = {
            'red': 0,
            'green': 0,
            'blue': 0
        }

        game, set = line.split(":")
        set = [s.strip() for s in set.replace(";", ",").split(",")] # convert sets of game into list of cubes
        
        # enumerate each cubes
        for s in set:
            num, cube = s.split(" ")
            if(int(num) > cubes[cube]):
                cubes[cube] = int(num)

        sum += cubes['red'] * cubes['green'] * cubes['blue']

    print(f'[+] part-2 solution: {sum}')

if __name__ == '__main__':
    sln_one()
    sln_two()